File: CodeTest/Python/Programmers/추석트래픽.py
def solution(lines):
    answer = 0                                  # 초당 최대 처리량
    cnt = 0                                     # 특정 시간대 처리량
    time_line = {}                              # 시간 테이블 - 키: 시간(ms) / 값: 시작 또는 완료 여부 (-1, 1)

    for line in lines:                          # 로그 데이터
        date, time, duration = line.split(' ')  # 날짜, 시간, 처리 시간
        end_ms = int(time[:2]) * 3600000 + int(time[3:5]) * 60000 + int(time[6:8]) * 1000 + int(time[9:])       # 요청 시간 (ms)
        start_ms = end_ms - int(duration[0]) * 1000 - (int(duration[2:-1]) if len(duration) > 2 else 0) + 1     # 완료 시간 (ms)
        end_ms += 1000                                          # 1초간 최대 처리량을 구해야 하므로 완료 시간을 1초 추가

        time_line[start_ms] = time_line.get(start_ms, 0) + 1    # 시작 시간을 키로 해서 값은 1
        time_line[end_ms] = time_line.get(end_ms, 0) - 1        # 완료 시간을 키로 해서 값은 -1

    for t in sorted(time_line.keys()):      # 키값들을 오름차순 정렬하여 시간 순서대로 반복
        cnt += time_line[t]                 # 값들 누적시키기
        answer = max(answer, cnt)           # 최대 처리량 구하기

    return answer


print(solution([
"2016-09-15 01:00:04.001 2.0s",
"2016-09-15 01:00:07.000 2s"
]))
print(solution([
"2016-09-15 01:00:04.002 2.0s",
"2016-09-15 01:00:07.000 2s"
]))
print(solution([
"2016-09-15 20:59:57.421 0.351s",
"2016-09-15 20:59:58.233 1.181s",
"2016-09-15 20:59:58.299 0.8s",
"2016-09-15 20:59:58.688 1.041s",
"2016-09-15 20:59:59.591 1.412s",
"2016-09-15 21:00:00.464 1.466s",
"2016-09-15 21:00:00.741 1.581s",
"2016-09-15 21:00:00.748 2.31s",
"2016-09-15 21:00:00.966 0.381s",
"2016-09-15 21:00:02.066 2.62s"
]))

File: CodeTest/JavaScript/programmers/추석 트래픽.js
function solution(lines) {
  let answer = 0          // 정답
  const timeline = {}     // 타임라인을 저장할 객체
  
  lines.map(v => v.split(' ')).forEach(v => {
      const hour = parseInt(v[1].slice(0, 2)) * 3600000       // 시간을 밀리초로 변경
      const minute = parseInt(v[1].slice(3, 5)) * 60000       // 분을 밀리초로 변경
      const second = parseFloat(v[1].slice(6)) * 1000         // 초를 밀리초로 변경
      
      let end = hour + minute + second                                            // 종료 시간
      const duration = parseFloat(v[2].replace('s', '')) * 1000                   // 소요 시간
      const start = end - duration + 1                                            // 시작 시간
      end += 1000                                                                 // 1초 동안의 처리량을 구해야하므로 종료 시간에 1초를 추가
      
      timeline[start] = timeline[start] === undefined ? 1 : timeline[start] + 1   // 시작 시간을 키값으로 값은 1
      timeline[end] = timeline[end] === undefined ? -1 : timeline[end] - 1        // 종료 시간을 키값으로 값은 -1
  })
  
  let cnt = 0
  Object.keys(timeline).sort((a,b) => a - b).forEach(v => {                       // 타임라인에 있는 시간들을 오름차순으로 정렬하여 1초동안 처리 가능한 최대 개수 구하기
      cnt += timeline[v] 
      answer = Math.max(answer, cnt)
  })
  
  return answer
}